### PR TITLE
Add dynamic native modules (MVP 0)

### PR DESCRIPTION
**Pull request by gbryant-arm**
Implement dynamic native modules based on https://github.com/veracruz-project/veracruz/discussions/577
This is the first step of an MVP.

### Done
* Add extension to policy file to specify name, special file (VFS file that interfaces with the native module) and entry point (path to main binary) per native module. For this MVP we're not looking at implementing a full fledged packaging system with checksums, signatures, links to artifacts, etc. The artifacts will have to be injected into the execution environment before deployment
* Replace legacy native module system with new system supporting dynamic native modules
* Write [sandboxer](https://github.com/veracruz-project/native-module-manager) (sandbox2 client)
* Implement native module management in the RM (sandbox environment init & teardown, native module launch)
* Unify static and dynamic NM

### Todo
* Polish

### Usage
Since no packaging system has been implemented yet, some binaries have to be injected into the execution environment before execution:
* Compile Veracruz, the [native module sandboxer](https://github.com/veracruz-project/native-module-sandboxer) and the [TF Lite native module](https://github.com/veracruz-project/tflite-nm)
* Create the `/tmp/nmm` directory in the execution environment (the machine/container where the Runtime Manager is to be executed)
* Copy `native-module-sandboxer` to `/tmp/nmm/`
* Copy `tflite-nm` to `/tmp/nmm/tflite-nm/`
* Deploy Veracruz

Note:
Dynamic native modules are expected to read the execution configuration file (default `/execution_config` within the sandbox) in order to know what to do, where to get inputs and where to write outputs.

### CODE REVIEW

**Comment by gbryant-arm** [comment]
RFR.
I'll run rustfmt and commit the lock files when I get approvals

File: execution-engine/src/fs.rs [hunk 1]
```diff
@@ -155,17 +153,11 @@ impl InodeEntry {
         self.data.is_service()
     }
 
-    /// Try to parse the input.
-    #[inline]
-    pub(crate) fn service_valid_input(&self) -> FileSystemResult<bool> {
-        self.data.service_valid_input()
-    }
```
[the remark this answers is not in the document]
Comment (gbryant-arm) [reply execution-engine/src/fs.rs:162]: `service_valid_input()` doesn't do much besides calling `try_parse()` defined in each static native module, so I figured it would be simpler to directly call `try_parse()` in the native module manager

File: execution-engine/src/fs.rs [hunk 2]
```diff
@@ -785,24 +769,20 @@ impl FileSystem {
...
     /// Wasm programs are going to expect that this is true, so we need to
     /// preallocate some files corresponding to those, here.
-    pub fn new(rights_table: RightsTable) -> FileSystemResult<Self> {
+    pub fn new(rights_table: RightsTable, native_modules: Vec<NativeModule>) -> FileSystemResult<Self> {
         let mut rst = Self {
             fd_table: HashMap::new(),
             next_fd_candidate: Self::FIRST_FD,
             inode_table: Arc::new(Mutex::new(InodeTable::new(rights_table)?)),
             prestat_table: HashMap::new(),
+            native_modules: native_modules.to_vec(),
```
[the remark this answers is not in the document]
Comment (gbryant-arm) [reply execution-engine/src/fs.rs:785]: I believe not using to_vec() here triggers the borrow checker. Let me try again

File: execution-engine/src/fs.rs [hunk 3]
```diff
@@ -1961,6 +1937,82 @@ impl FileSystem {
...
+        let mut rst = Vec::new();
+        let mut top_level_files = Vec::new();
+        if self.lock_inode_table()?.is_dir(&inode) {
+            // Limit the lock scope
+            let all_dir = {
+                let inode_table = self.lock_inode_table()?;
+                inode_table.get(&inode)?.read_dir(&inode_table)?
+            };
+            let iter = all_dir.iter();
+            if iter.count() <= 2 {
```
[the remark this answers is not in the document]
Comment (gbryant-arm) [reply execution-engine/src/fs.rs:1977]: For improved readability we can implement an `is_dir_empty()` function that performs this check, if that's what you mean

File: execution-engine/src/fs.rs [hunk 4]
```diff
@@ -2057,6 +2109,6 @@ where
     }
 }
 
-fn strip_root_slash(path: &Path) -> &Path {
+pub fn strip_root_slash(path: &Path) -> &Path {
```
[the remark this answers is not in the document]
Comment (gbryant-arm) [reply execution-engine/src/fs.rs:2112]: The native module manager needs `strip_root_slash()`, which is currently defined in `fs.rs`.
`pub crate` is more restrictive and should work.

File: execution-engine/src/native_module_manager.rs [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+        }
+
+        Ok(())
+    }
+}
+
+impl Drop for NativeModuleManager {
+    /// Drop the native module manager.
+    fn drop(&mut self) {
+        //let _ = self.teardown_fs();
```
[the remark this answers is not in the document]
Comment (gbryant-arm) [reply execution-engine/src/native_module_manager.rs:291]: I could enable `teardown_fs()` back, though it currently doesn't add any security, since native modules and program have the same permissions. It will be more useful when native modules only have access to a subset of the program's VFS

File: execution-engine/src/fs.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (dreemkiller) [reply execution-engine/src/fs.rs:1977]: I think that would be clearer.

File: execution-engine/src/fs.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dreemkiller) [reply execution-engine/src/fs.rs:785]: Any result on that?

**Review by dreemkiller — CHANGES_REQUESTED** [review]
Did you take care of the `Cargo.lock` files? I approve, but need to make sure the `Cargo.lock` files are done according to @mathias-arm 's process.

**Comment by gbryant-arm** [comment]
> Did you take care of the `Cargo.lock` files? I approve, but need to make sure the `Cargo.lock` files are done according to @mathias-arm 's process.

I've just checked, the Linux, IceCap and Nitro runtime lock files are committed :heavy_check_mark:

File: execution-engine/src/native_module_manager.rs [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+                    let mut f = File::open(self.native_module_directory.join(&path_prefixed))?;
+                    let mut buf: [u8; 128] = [0; 128];
+
+                    // Copy it to the VFS. First truncate the VFS file first
+                    // then append to it. If the principal doesn't have write
+                    // access, just ignore it
+                    if self.native_module_vfs.write_file_by_absolute_path(&path_unprefixed, vec![], false).is_ok() {
+                        loop {
+                            let n = f.read(&mut buf)?;
+                            if n <= 0 {
```
Comment (ShaleXIONG) [inline execution-engine/src/native_module_manager.rs:200]: Minor: `n` is of type`usize` so cannot be negative.

Another way to write this without `break` command is `while f.read(&mut buf)? > 0 { ... }`

File: execution-engine/src/native_module_manager.rs [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+
+    /// Run the native module. The input is passed by the WASM program via the
+    /// native module's special file.
+    pub fn execute(&mut self, input: Vec<u8>) -> FileSystemResult<()> {
+        if self.native_module.is_static() {
+            // Look up native module in the static native modules table
+            let mut nm = STATIC_NATIVE_MODULES
+                .lock()
+                .map_err(|_| ErrNo::Inval)?;
+            let nm = nm
```
Comment (ShaleXIONG) [inline execution-engine/src/native_module_manager.rs:231]: Minor: can chain together with the previous `let` command.

File: execution-engine/src/native_module_manager.rs [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+                    "--sandbox2tool_resolve_and_add_libraries",
+                    "--sandbox2tool_mount_tmp",
+                    "--sandbox2tool_additional_bind_mounts",
+                    &mount_mappings,
+                    &self.native_module.entry_point_path().to_str().ok_or(ErrNo::Inval)?.to_owned(),
+                ])
+                .output()?;
+
+            info!("Propagating side effects to the VFS...");
+            self.copy_fs_to_vfs(&PathBuf::from(""))?;
```
Comment (ShaleXIONG) [inline execution-engine/src/native_module_manager.rs:276]: Why does this line trigger the propagation of side effects to VFS?

**Review by ShaleXIONG — COMMENTED** [review]
Feel free to ignore the minor suggestions. Can you clarify one question I left? thanks.

File: execution-engine/src/native_module_manager.rs (hunk 7 above)
Comment (gbryant-arm) [reply execution-engine/src/native_module_manager.rs:231]: Chaining `let` creates a temporary which is freed while still in use

File: execution-engine/src/native_module_manager.rs (hunk 6 above)
Comment (gbryant-arm) [reply execution-engine/src/native_module_manager.rs:200]: We actually need to access `n` inside the loop

File: execution-engine/src/native_module_manager.rs (hunk 8 above)
Comment (gbryant-arm) [reply execution-engine/src/native_module_manager.rs:276]: Solved offline

File: execution-engine/src/native_module_manager.rs [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+    /// Build kernel-to-sandbox filesystem mappings palatable to the native
+    /// module sandboxer.
+    /// Takes a list of unprefixed paths, i.e. not including the path to the
+    /// native module's directory.
+    /// Returns the mappings as a string.
+    fn build_mappings(&self, unprefixed_files: Vec<PathBuf>) -> FileSystemResult<String> {
+        let mut mappings = String::new();
+        for f in unprefixed_files {
+            let mapping = self.native_module_directory.join(strip_root_slash(&f));
+            let mapping = mapping.to_str().ok_or(ErrNo::Inval)?.to_owned()
```
[the remark this answers is not in the document]
Comment (dreemkiller) [reply execution-engine/src/native_module_manager.rs]: @gbryant-arm <opinion> Please don't do that. Long lines of chained commands hurt readability</opinion>

[the remark this answers is not in the document]
Comment (gbryant-arm) [reply execution-engine/src/native_module_manager.rs]: @dreemkiller Whoops. I can reverse if you like, I don't mind either

**Review by ShaleXIONG — APPROVED** [review]

**Review by dreemkiller — APPROVED** [review]
